### ISSUE TITLE
Bug : Left Hand Nav (LHN) populated at top with many zero dollar balances or "$0 balances not being removed so they're showing in LHN"

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#11596]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Opened app after sleeping for hour

## Expected Result:
LHN to be the same as when it went to sleep and with new chats

## Actual Result:
LHN now shows many red and green bars next to new chats at top for amounts owed to me and others

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.11-3
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
<img width="374" alt="2022-10-04_11-22-03" src="https://user-images.githubusercontent.com/43996225/193938895-08793ae2-3bbc-4d6e-8876-a45673ed90d1.png">
<img width="954" alt="Untitled1" src="https://user-images.githubusercontent.com/43996225/193938915-845eb8c9-1bcf-49f0-9bf6-ee4e079e6488.png">

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @mallenexpensify 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1664918444481379

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by AndrewGable** [comment]
I have a suspicion this was caused by the web deploy because the timing seems to line up, but cannot point to anything that confirms it yet.

**Comment by danieldoglas** [comment]
cc @Julesssss @mountiny

**Comment by cead22** [comment]
I wasn't able to reproduce

**Comment by cead22** [comment]
Never mind, I was able to reproduce. It seems like you simply need to sign into an account that has past P2P payments, and it's only happening on staging (I couldn't repro on prod)

**Comment by thienlnam** [comment]
Yeah, I am able to reproduce this on my staging with my account that has previous P2P payments like Carlos mentioned.

https://github.com/Expensify/App/blob/1a42cf8a48bcb51b7b469f342a3f122de7aaef34/src/components/LHNOptionsList/OptionRowLHN.js#L207-L209

> I have a suspicion this was caused by the web deploy because the timing seems to line up, but cannot point to anything that confirms it yet.

I also agree, we haven't deployed App since and this bug only showed up for me today at least

Looks like it was added from this PR https://github.com/Expensify/Web-Expensify/pull/34924/files

**Comment by thienlnam** [comment]
Yeah this looks like it is not longer a bug but from old IOUs. I've been testing this and requesting something / canceling it will resolve the issue.  I believe this was in the past when you would request an amount and it would be declined or someone would create an IOU that would offset the amount and it did not resolve

**Comment by thienlnam** [comment]
When you check the linked IOU https://github.com/Expensify/Web-Expensify/blob/648d564eaebb12e670eda50e9a3cdc09395506f5/lib/ReportUtils.php#L1827-L1827, it has hasOutstandingIOU set to false if the amount is 0.

However, when we build the reportData for onyx [here](https://github.com/Expensify/Web-Expensify/blob/648d564eaebb12e670eda50e9a3cdc09395506f5/lib/ReportUtils.php#L2211-L2247), we only check if an IOU is linked.  I'm not sure why we have hasOutstandingIOU keys in different spots and have them represent different things do you guys have extra context? cc @Julesssss @mountiny @Justicea83 

There seems to be a few different solutions here
1. When we decline an IOU request we update Onyx to remove the linked IOU to the report or change the state of the IOU (which looks like it might be done in the IOU refactor here? https://github.com/Expensify/Expensify/issues/213615)

2. Update App to check hasOutstandingIOU of the linked IOU report in SidebarLinks.js

3.  Quick fix to update ReportUtils::structureReportForOnyx https://github.com/Expensify/Web-Expensify/blob/648d564eaebb12e670eda50e9a3cdc09395506f5/lib/AppInit.php#L120-L132 to fetch an active IOU report and see if it's value is 0

**Comment by thienlnam** [comment]
I'm going to push up a Web-E PR for 3 to clear the blocker, and then when we finish the refactors for IOUs then we'll do solution 1 which is a better solution for when we decline / cancel transactions

Not going to revert the PR https://github.com/Expensify/Web-Expensify/pull/34924 because that actually fixed another regression that was added and so I'll push up a patch PR

**Comment by mountiny** [comment]
Thanks for the investigation! Havent experienced this issue when testing locally, also a quick dirty fix could be in App (which is probably worse than in Web-E) to not show the badge if the value is `0.00`.

**Comment by Julesssss** [comment]
I know what causes this and we have a couple of ways of fixing it. The real fix is to update the Auth command so that it returns iouReports, but a quick fix can be applied to App also. I will work on this and submit a PR today.

As to why this regression occurred, this was thoroughly tested against an up to date Auth `main` build, so I'm pretty confused as to why this has different behavior on staging. Notice that the PR tests against this exact case 😕

**Comment by iwiznia** [comment]
I am very confused by this bug. I checked one of my reports and it seems that GetChats for report 65756440 is returning:
```
hasIOUAction":true,"iouReportID":98351486
```

And that IOU has $0 amount, but we don't check that [here](https://github.com/Expensify/Auth/blob/93824370af8f5c9eebf7e56c2f2a5fc8c2c8c64b/auth/lib/Report.cpp#L3253-L3263)... so I think that logic is wrong and should be updated, since there's nothing outstanding.
I have no idea why/how this started to show up now, but almost sure that the fix should be in auth.

Full output of my queries:
```
ionatan@db2.sjc:~$ nc localhost 4444
GetChats
authToken:X
reportIDList:65756440

200 OK
commitCount: 17184166940
nodeName: auth.db2.sjc
peekTime: 66601
requestID: A0X7Sp
totalTime: 66744
unaccountedTime: 143
Content-Length: 1605

{"accountID":2536717,"authToken":"X","email":"ionatan@expensify.com","reportList":[{"hasIOUAction":true,"iouReportID":98351486,"isOwnPolicyExpenseChat":false,"isPinned":false,"lastActionActorEmail":"ionatan@expensify.com","lastActionCreated":"2022-08-05 19:10:13","lastActionMessage":{"clientID":"1659726612349806","html":":-) you too!"},"lastActionType":"ADDCOMMENT","managerEmail":"__fake__","ownerEmail":"__fake__","participants":["kevin@expensify.com"],"reportActionCount":1807,"reportID":65756440,"reportName":"Chat Report","reportNameValuePairs":{"expensify_policyID":"_FAKE_","lastNotifiedActionID":"1807","lastReadActionID_0":"39","lastRead_2536717":{"sequenceNumber":1807,"timestamp":1664970855007},"lastRead_4087623":{"sequenceNumber":1807,"timestamp":1659726671944},"type":"chat"},"sharedReportList":[{"accountID":2536717,"email":"ionatan@expensify.com","permissions":"read, write"},{"accountID":4087623,"email":"kevin@expensify.com","permissions":"read, write"}],"state":0,"status":0}]}^C
ionatan@db2.sjc:~$ getreport.sh 98351486
Report 98351486 is Submitted
State: 1: Submitted Status: 1: Submitted
Report:
created,submitted,approved,reportID,accountID,managerID,reportName,state,total,currency,status
"2022-08-01 22:10:34","2022-08-01 22:10:34",,98351486,2536717,4087623,IOU,1,0,EUR,1

Submitter:
accountID,email
2536717,ionatan@expensify.com

SharedReports:
accountID,reportID,permissions
4087623,98351486,"read, write"

ReportActions:
created,reportID,accountID,action,message,reportActionID
"2022-08-01 22:10:34",98351486,2536717,CREATED,"",1082165213
"2022-08-01 22:10:39",98351486,2536717,BILLABLEUPDATETRANSACTION,"{""action"":""unreport"",""transactionID"":13016680074577139}",1082165264
"2022-08-01 22:10:39",98351486,2536717,REJECTEDTRANSACTION,"{""created"":""2022-08-01 22:10:34"",""merchant"":""IOU"",""amount"":""(CR₡23.00)"",""message"":""""}",1082165268

ReportNVPs:
reportID,name,value
98351486,expensify_policyID,1DEFA6D665102BF4
98351486,type,iou
```

**Comment by iwiznia** [comment]
Oh, just noticed that this is on staging but not prod, so the bug is triggered by a change in newDot, but still I think the backend is not doing what it should.

**Comment by Julesssss** [comment]
Yep, the problem is in Auth. I was working on it [here](https://github.com/Expensify/Auth/pull/7095) and and fixed the issue, but a better solution is to [start retrieving iouReports](https://github.com/Expensify/Expensify/issues/231746) as part of the GetChats call.

**Comment by iwiznia** [comment]
That's not a solution, that's a performance improvement. We still need to not return the IOU if the amount is 0

**Comment by Julesssss** [comment]
The [first link](https://github.com/Expensify/Auth/pull/7095) was trying to resolve this in Auth. 

The PR 100% worked btw, but I couldn't figure out why this test was failing. Does anyone have an idea? If so we could resolve this immediately.

**Comment by Julesssss** [comment]
[Here's](https://github.com/Expensify/Auth/pull/7095#issuecomment-1268714205) the best theory as for why the test is failing. If we can resolve this, then we can merge the PR and fix the issue.

**Comment by AndrewGable** [comment]
Per discussion [in Slack here ](https://expensify.slack.com/archives/C07J32337/p1664992277722249?thread_ts=1664903854.322709&cid=C07J32337) we pinpointed this issue to this API PR: https://github.com/Expensify/Web-Expensify/pull/34924

Since this isn't going to effect production NewDot, it's not a deploy blocker and we will fix the API PR here: https://github.com/Expensify/Web-Expensify/pull/35059

**Comment by Julesssss** [comment]
Closing this as the original PR was reverted.
